### ISSUE TITLE
LoadLibrary failure for Prophet on Windows

### ISSUE THREAD

**Issue by seanjtaylor** [stan-dev/rstantools#81]
As mentioned in #76 we're unable to install Prophet 0.7 right now on Windows machines. I tested on win-builder and then replicated on my local machine. See errors below. I'd love to figure out what's going on here so we can get this on CRAN.

A full error log from win-builder is [gisted here](https://gist.github.com/seanjtaylor/65a39dde3fe92f4ccbc63ab7f86be57b).

```
** testing if installed package can be loaded from temporary location
*** arch - i386
Error: package or namespace load failed for 'prophet' in inDL(x, as.logical(local), as.logical(now), ...):
 unable to load shared object 'C:/Users/Sean/Documents/R/win-library/4.0/00LOCK-prophet/00new/prophet/libs/i386/prophet.dll':
  LoadLibrary failure:  The specified module could not be found.

Error: loading failed
Execution halted
*** arch - x64
Error: package or namespace load failed for 'prophet' in inDL(x, as.logical(local), as.logical(now), ...):
 unable to load shared object 'C:/Users/Sean/Documents/R/win-library/4.0/00LOCK-prophet/00new/prophet/libs/x64/prophet.dll':
  LoadLibrary failure:  %1 is not a valid Win32 application.

Error: loading failed
Execution halted
ERROR: loading failed for 'i386', 'x64'
* removing 'C:/Users/Sean/Documents/R/win-library/4.0/prophet'
Error: Failed to install 'prophet' from GitHub:
  (converted from warning) installation of package ‘C:/Users/Sean/AppData/Local/Temp/RtmpiSgvDs/file53062fd7c7a/prophet_0.7.tar.gz’ 
```

**Comment by jgabry** [comment]
Hmm, yeah that's frustrating, sorry! I'm definitely not the resident windows expert here so I'm not sure what's going on. @bgoodri @wds @rok-cesnovar or anyone else who has/knows Windows, have you seen this before?

**Comment by wds15** [comment]
I probably should have posted this here:

I did not find any more the thread with the prophet issue… anyway. It looks as if the prophet R package misses in the DESCRIPTION file to import RcppParallel and they als do not list this under “LinkingTo:”. I hope this fixes the issues under Windows. If you ping me on the right github issue then we can dig further as needed.

EDIT: I can confirm that prophet 0.7 now loads for me under Windows just fine with the changes as outlined. So we should have a check in the rstantools config thing which ensure that RcppParallel is in the imports and linkingTo section of the descirption file.

In short: The prophet is missing important fields in the DESCRIPTION file. It works for me on windows.

@jgabry I would recommend that the rstan_config() function reads in and checks the DESCRIPTION file of the respective package calling the routine. The RcppParallel package must be listed under imports and under linkingto. If this is not the case, then the routine should error with an adequate message saying what needs to be corrected.

**Comment by jgabry** [comment]
Awesome, thanks @wds15 for figuring this out! And yeah your recommendation for `rstan_config()` sounds like a good idea.

**Comment by bgoodri** [comment]
It is possible that we can do in src/Makevars{.win} what LinkingTo:
RcppParallel does in DESCRIPTION.

**Comment by jgabry** [comment]
@bgoodri that's fine by me if you can figure out how to get that working.

**Comment by seanjtaylor** [comment]
@wds15 can you share the modifications you made to the `DESCRIPTION` file? I wasn't able to replicate your successful loading on win-builder after adding to the `Imports` and `LinkingTo` blocks.

**Comment by wds15** [comment]
Could you post your `DESCRIPTION` file?

**Comment by seanjtaylor** [comment]
I also bumped up version numbers on `rstan`:

```
Package: prophet
Title: Automatic Forecasting Procedure
Version: 0.7
Date: 2020-10-18
Authors@R: c(
  person("Sean", "Taylor", email = "sjtz@pm.me", role = c("cre", "aut")),
  person("Ben", "Letham", email = "bletham@fb.com", role = "aut")
  )
Description: Implements a procedure for forecasting time series data based on
    an additive model where non-linear trends are fit with yearly, weekly, and
    daily seasonality, plus holiday effects. It works best with time series
    that have strong seasonal effects and several seasons of historical data.
    Prophet is robust to missing data and shifts in the trend, and typically
    handles outliers well.
URL: https://github.com/facebook/prophet
BugReports: https://github.com/facebook/prophet/issues
Depends:
    R (>= 3.4.0),
    Rcpp (>= 0.12.0),
    rlang (>= 0.3.0.1)
Imports:
    dplyr (>= 0.7.7),
    dygraphs (>= 1.1.1.4),
    extraDistr,
    ggplot2,
    grid,
    methods,
    RcppParallel (>= 5.0.1),
    rstan (>= 2.21.2),
    rstantools (>= 2.0.0),
    scales,
    StanHeaders,
    stats,
    tidyr (>= 0.6.1),
    xts
Suggests:
    knitr,
    testthat,
    readr,
    rmarkdown
SystemRequirements: GNU make, C++11
Biarch: true
License: MIT + file LICENSE
LinkingTo: 
    BH (>= 1.66.0),
    Rcpp (>= 0.12.0),
    RcppEigen (>= 0.3.3.3.0),
    RcppParallel (>= 5.0.1),
    rstan (>= 2.21.2),
    StanHeaders (>= 2.21.0)
LazyData: true
VignetteBuilder: knitr
Encoding: UTF-8
RoxygenNote: 7.1.1
```

**Comment by wds15** [comment]
Looks actually ok to me. Here is what installed into my virtual windows machine:

```
Package: prophet
Title: Automatic Forecasting Procedure
Version: 0.7
Date: 2020-02-10
Authors@R: c(
  person("Sean", "Taylor", email = "sjtz@pm.me", role = c("cre", "aut")),
  person("Ben", "Letham", email = "bletham@fb.com", role = "aut")
  )
Description: Implements a procedure for forecasting time series data based on
    an additive model where non-linear trends are fit with yearly, weekly, and
    daily seasonality, plus holiday effects. It works best with time series
    that have strong seasonal effects and several seasons of historical data.
    Prophet is robust to missing data and shifts in the trend, and typically
    handles outliers well.
URL: https://github.com/facebook/prophet
BugReports: https://github.com/facebook/prophet/issues
Depends:
    R (>= 3.4.0),
    Rcpp (>= 0.12.0),
    rlang (>= 0.3.0.1)
Imports:
    dplyr (>= 0.7.7),
    dygraphs (>= 1.1.1.4),
    extraDistr,
    ggplot2,
    grid,
    methods,
    rstan (>= 2.18.1),
    rstantools (>= 2.0.0),
    scales,
    StanHeaders,
    stats,
    tidyr (>= 0.6.1),
    xts,
    RcppParallel (>= 5.0.1)
Suggests:
    knitr,
    testthat,
    readr,
    rmarkdown
SystemRequirements: GNU make, C++11
Biarch: true
License: MIT + file LICENSE
LinkingTo:
    BH (>= 1.66.0),
    Rcpp (>= 0.12.0),
    RcppEigen (>= 0.3.3.3.0),
    rstan (>= 2.18.1),
    StanHeaders (>= 2.18.0),
    RcppParallel (>= 5.0.1)
LazyData: true
VignetteBuilder: knitr
Encoding: UTF-8
RoxygenNote: 7.1.1
```

... and I also edited the NAMESPACE to be

```
# Generated by roxygen2: do not edit by hand

S3method(plot,prophet)
S3method(predict,prophet)
export(add_changepoints_to_plot)
export(add_country_holidays)
export(add_regressor)
export(add_seasonality)
export(cross_validation)
export(dyplot.prophet)
export(fit.prophet)
export(make_future_dataframe)
export(performance_metrics)
export(plot_cross_validation_metric)
export(plot_forecast_component)
export(predictive_samples)
export(prophet)
export(prophet_plot_components)
import(Rcpp)
import(RcppParallel)
import(rlang)
importFrom(dplyr,"%>%")
useDynLib(prophet, .registration = TRUE)
```

I am not so sure whats wrong with you DESCRIPTION... maybe it really also needs the NAMESPACE updates which I obviously forgot. Does it work now?

**Comment by wds15** [comment]
This is fixed after all and can be closed I suppose.

**Comment by andrjohns** [comment]
Closing as resolved
